bcachefs: 2023-06-28 -> 2023-09-22

YellowOnion opened the pull request:
This release includes an overhaul of various bcachefs systems in prep for mainline merge.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

@Madouura @Mic92 @davidak @pedrohlc @raitobezarius

YellowOnion commented:
This supersedes: https://github.com/NixOS/nixpkgs/pull/252348

RaitoBezarius commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29029,7 +29029,11 @@ with pkgs;
   };
-
+  # remove when util-linux 2.39.3 comes out
+  systemdBcachefs = systemd.override {
```
Shouldn't we just pull the util-linux patch? Not so sure about a new systemd…

YellowOnion replied:
@RaitoBezarius I tried applying the patches but it wouldn't compile, I opened a bug report to get a stable release with patches applied, there's about 5 or so patches in master.

systemd hardcodes the mount paths to `util-linuxMinimal` it took me ages to figure out why `systemd-remount-fs` service would fail but everything else wouldn't.  this is the solution to that.

I was trying to avoid using util-linux from master on non-bcachefs nixpkgs, if you can figure out how to get util-linux to compile with just the minimal set of patches we can go with that.

RaitoBezarius replied:
IMHO, this is a blocker because I already have patches on the top of systemd and this will make things much more complicated for developers like me who works on systemd.

We should find a way to put those patches inside the default util-linuxMinimal.

YellowOnion commented on nixos/modules/tasks/filesystems/bcachefs.nix:
```diff
@@ -51,28 +42,29 @@ in
         copy_bin_and_libs ${pkgs.bcachefs-tools}/bin/bcachefs
-        copy_bin_and_libs ${mountCommand}/bin/mount.bcachefs
+        substitute ${pkgs.bcachefs-tools}/bin/mount.bcachefs $out/bin/mount.bcachefs \
+          --replace '${pkgs.bash}/bin/sh' /bin/sh                                    \
```
I just found that this assumes that `mount.bcachefs` uses the bash from `pkgs`, which isn't always the case if we're using an overlay from a different source (i.e. using bcachefs from unstable, and system from stable)

YellowOnion committed:
bcachefs: 2023-06-28 -> 2023-09-29

Includes prep work for mainline bcachefs release.

Mic92 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29014,7 +29014,11 @@ with pkgs;
   };
-
+  # remove when util-linux 2.39.3 comes out
+  systemdBcachefs = systemd.override {
```
nitpick: I think even for overrides we can already use pkgs/by-name

YellowOnion replied:
@Mic92 Yeah, I just did a quick and dirty change until the main util-linux gets a patch. I was considering moving tools in to by-name.

Mic92 commented on pkgs/os-specific/linux/kernel/update.sh:
```diff
@@ -12,3 +12,6 @@ COMMIT=1 ./update-libre.sh || echo "update-libre failed with exit code $?"
 echo "Update linux-hardened"
 COMMIT=1 ./hardened/update.py || echo "update-hardened failed with exit code $?"
+
+echo Update linux-testing-bcacehfs
```
How long do you think we still have this kernel? If the mainline merge goes through this would be just a matter of weeks?

YellowOnion replied:
I just added it because it made my job simpler while writing this patch, I already had one sitting around, I assume bcachefs is coming in v6.7 which is a bit longer time schedule.

We can pull it out in to a `nur-kernels` as @RaitoBezarius suggested in the IRC.

RaitoBezarius replied:
6.7 will be here in November FWIW.

RaitoBezarius replied:
The issue now is that we are dropping 6.4 in #258569, so actually, it would be good to extract the meaningful part to bump to 6.5 and get it ready ASAP even before November.

PedroHLC reviewed (approved):
No issues for me as long as we remember to drop all the workaround when we drop the "testing" kernel.

TL;DR: LGTM.

Madouura commented:
@YellowOnion
```console
nix build --print-out-paths -L .#bcachefs-tools.tests.bcachefsSimple
trace: warning: mdadm: Neither MAILADDR nor PROGRAM has been set. This will cause the `mdmon` service to crash.
bcachefs> trying https://evilpiepirate.org/git/bcachefs.git/rawdiff/?id=508de596ed0b58b1dbe783ca1fbbee04d40e9ccf&id2=v6.5
bcachefs>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
bcachefs>                                  Dload  Upload   Total   Spent    Left  Speed
console-env> created 333 symlinks in user environment
bcachefs> 100 2611k    0 2611k    0     0  3003k      0 --:--:-- --:--:-- --:--:-- 3001k
fontconfig-etc> created 1 symlinks in user environment
error: hash mismatch in fixed-output derivation '/nix/store/ghrvp7r3lynprk4dk9iaxhkhmrk2z5lj-bcachefs-508de596ed0b58b1dbe783ca1fbbee04d40e9ccf.diff.drv':
         specified: sha256-wNMttW681D47YYXhRXjRT4/QWdFM0pOcKrrtZvgyd5w=
            got:    sha256-MdSoXJVxytSbIoJr73vD5gr9UC5Hwh4n6i28ElSeOUE=
error: 1 dependencies of derivation '/nix/store/kjvnkj2xaqgw6llk01jajm6svckfvvrn-linux-6.5.4.drv' failed to build
error: 1 dependencies of derivation '/nix/store/la77m480b0dvr4a8ld9vqm016224g2aj-linux-6.5.4-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/mvi3gahq31ipl2vvlmrac7wiz285w4ln-nixos-system-machine-23.11pre-git.drv' failed to build
error: 1 dependencies of derivation '/nix/store/j02a7i9mh2b194j50l7si39yrm0hprnr-nixos-vm.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ryrflimbb912wxkhzwsmk8r117nymjmg-nixos-test-driver-installer-bcachefs-simple.drv' failed to build
error: 1 dependencies of derivation '/nix/store/kgmjvjjiajpiapnp9zyxgxnlnf8pcqrb-vm-test-run-installer-bcachefs-simple.drv' failed to build
```
Replacing the hash fixes it.

Madouura commented:
All tests except for `bcachefsEncrypted` work. Same error as last time.
```console
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    0.970516] virtio_blk virtio3: [vda] 16777216 512-byte logical blocks (8.59 GB/8.00 GiB)
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    0.979977]  vda: vda1 vda2 vda3
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    0.982396] sr 1:0:0:0: [sr0] scsi3-mmc drive: 4x/4x cd/rw xa/form2 tray
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    0.983363] cdrom: Uniform CD-ROM driver Revision: 3.20
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.154878] raid6: avx2x4   gen() 36054 MB/s
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.171877] raid6: avx2x2   gen() 30172 MB/s
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.188878] raid6: avx2x1   gen() 28696 MB/s
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.189472] raid6: using algorithm avx2x4 gen() 36054 MB/s
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.206877] raid6: .... xor() 6476 MB/s, rmw enabled
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.207549] raid6: using avx2x2 recovery algorithm
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.211393] xor: automatically using best checksumming function   avx
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.375898] I/O error, dev fd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 2
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.377013] floppy: error 10 while reading block 0
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.382647] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.383690] CPU: 0 PID: 1 Comm: init Not tainted 6.5.4 #1-NixOS
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.384467] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.16.2-0-gea1b7a073390-prebuilt.qemu.org 04/01/2014
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.385925] Call Trace:
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.386320]  <TASK>
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.386667]  dump_stack_lvl+0x47/0x60
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.387196]  panic+0x323/0x340
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.387654]  ? srso_alias_return_thunk+0x5/0x7f
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.388279]  do_exit+0x959/0xad0
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.388800]  ? handle_mm_fault+0x9e/0x350
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.389375]  do_group_exit+0x31/0x80
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.389888]  __x64_sys_exit_group+0x18/0x20
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.390473]  do_syscall_64+0x3e/0x90
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.390989]  entry_SYSCALL_64_after_hwframe+0x6e/0xd8
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.391671] RIP: 0033:0x7f5b8c6d42ed
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.392186] Code: 01 48 83 c8 ff c3 0f 1f 80 00 00 00 00 f3 0f 1e fa 48 8b 35 1d 4b 10 00 ba e7 00 00 00 eb 07 66 0f 1f 44 00 00 f4 89 d0 0f 05 <48> 3d 00 f0 ff ff 76 f3 f7 d8 64 89 06 eb ec 0f 1f 40 00 f3 0f 1e
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.394522] RSP: 002b:00007ffd5fa06ae8 EFLAGS: 00000202 ORIG_RAX: 00000000000000e7
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.395542] RAX: ffffffffffffffda RBX: 00007ffd5fa07f04 RCX: 00007f5b8c6d42ed
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.396452] RDX: 00000000000000e7 RSI: ffffffffffffff88 RDI: 0000000000000001
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.397361] RBP: 0000000000000004 R08: 00007ffd5fa06af0 R09: 0000000000000004
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.398275] R10: 02b87a7c7497567c R11: 0000000000000202 R12: 0000000000000000
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.399190] R13: 00007ffd5fa06ed0 R14: 0000000000000000 R15: 00007f5b8c924000
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.400105]  </TASK>
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.400524] Kernel Offset: 0x36c00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
vm-test-run-installer-bcachefs-encrypted> boot-after-install # [    1.401912] Rebooting in 1 seconds..
```

Mic92 commented on pkgs/tools/filesystems/bcachefs-tools/update.sh:
```diff
@@ -0,0 +1,13 @@
+DATE=$(git --git-dir=bcachefs-tools.git log --date=short --pretty=format:%ad HEAD)
+rm -rf bcachefs-tools.git
+
+nix-prefetch-github koverstreet bcachefs-tools | jq --arg date $DATE '. + {date:$date}' > version.json
```
I might be biased, but you could achieve the same using `nix-update --version=branch` without using json and `updateScript = nix-update-script { extraArgs = [ "--version=branch" ]; };`. Might be less code (also you might have to pull from github). But now the code written anyway...

RaitoBezarius replied:
@Mic92 offtopic but do we have documentation on how to use `nix-update-script` in nixpkgs?

Mic92 commented:
We could greatly simplify the amount of packages by just add this patchset to utillinux: https://github.com/util-linux/util-linux/pull/2505

YellowOnion commented:
@Madouura I was meaning to run that test and got distracted, I know what the issue is, the device is encrypted so we can't find the UUID, I'm unsure how we're meant to handle this without also specifying a custom flag / parameter. Maybe I'll write a patch for upstream.

Madouura reviewed (approved):
Since the encryption issue in the VM test is an upstream issue, LGTM.
Might wanna update to https://github.com/koverstreet/bcachefs/commit/0a84850d91671606c6272fe4679e27dd68daf88e since apparently there was a lock leak.

YellowOnion commented:
@Madouura I found a work around for the encryption stuff, I can use `blkid` to get the device path, since bcachefs has a plain text super. I have a test branch in `bump-bcachefs-test` trying to figure out why cryptsetup isn't passing tests with the new `util-linux` I skipped the tests for now, but compile is taking a while.

YellowOnion committed:
bcachefs-tools: remove unnecessary dependency on valgrind

YellowOnion committed:
bcachefs-tools: reintroduce format-security error.

I believe this has been fixed upstream.

Madouura commented:
Testing with this diff on rebased upstream/master to allow for tracking the commit date.
Will let you know when it's done.
```diff
diff --git a/pkgs/os-specific/linux/kernel/bcachefs.json b/pkgs/os-specific/linux/kernel/bcachefs.json
index 852ca219efa2..09f7ea2febaa 100644
--- a/pkgs/os-specific/linux/kernel/bcachefs.json
+++ b/pkgs/os-specific/linux/kernel/bcachefs.json
@@ -1,4 +1,5 @@
 {
   "diffHash": "sha256-BmXd/5Hn/xr7gNFp6BsBMG2XeKFlPGxv66IQ8DEwh5k=",
-  "commit": "4d2faeb4fb58c389dc9f76b8d5ae991ef4497e04"
+  "commit": "4d2faeb4fb58c389dc9f76b8d5ae991ef4497e04",
+  "date": "2023-09-28"
 }
diff --git a/pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix b/pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix
index f4fc5361003d..18af0f55bd74 100644
--- a/pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix
+++ b/pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix
@@ -2,7 +2,6 @@
 , stdenv
 , fetchpatch
 , kernel
-, commitDate ? "2023-06-28"
 # bcachefs-tools stores the expected-revision in:
 #   https://evilpiepirate.org/git/bcachefs-tools.git/tree/.bcachefs_revision
 # but this does not means that it'll be the latest-compatible revision
@@ -18,7 +17,7 @@ in
 
 # NOTE: bcachefs-tools should be updated simultaneously to preserve compatibility
 (kernel.override ( args // {
-  version = "${kernel.version}-bcachefs-unstable-${commitDate}";
+  version = "${kernel.version}-bcachefs-unstable-${version.date}";
 
   extraMeta = {
     branch = "master";
diff --git a/pkgs/os-specific/linux/kernel/update-bchfs.sh b/pkgs/os-specific/linux/kernel/update-bchfs.sh
index 7d54c9274587..30135da9a2d3 100755
--- a/pkgs/os-specific/linux/kernel/update-bchfs.sh
+++ b/pkgs/os-specific/linux/kernel/update-bchfs.sh
@@ -11,4 +11,10 @@ COMMIT=$(git ls-remote $URL_BASE HEAD | awk '{ print $1; }')
 
 URL="$URL_BASE/rawdiff/?id=${COMMIT}&id2=v${VERSION}"
 
-nix store prefetch-file --name bcachefs-${COMMIT}.diff --json $URL |  jq --arg c "$COMMIT" '{diffHash: .hash, commit: $c}' > bcachefs.json
+rm -rf bcachefs.git
+git clone --depth=1 --bare $URL_BASE
+
+DATE=$(git --git-dir=bcachefs.git log --date=short --pretty=format:%ad HEAD)
+rm -rf bcachefs.git
+
+nix store prefetch-file --name bcachefs-${COMMIT}.diff --json $URL |  jq --arg c "$COMMIT" --arg date $DATE '{diffHash: .hash, commit: $c, date: $date}' > bcachefs.json

```

PedroHLC commented:
It's also probably feasible to enhance the update-script to use [their atom feeds](https://evilpiepirate.org/git/bcachefs.git/atom/?h=master) instead of fetching the git on every run.

But let's keep update-script suggestions in another PR to reduce the review size?

Madouura reviewed (changes requested):
I had to disable `cryptsetup` tests, since the `util-linux` changes broke them.
I am unsure if we can merge this since we are breaking this.

YellowOnion commented:
> I had to disable `cryptsetup` tests, since the `util-linux` changes broke them. I am unsure if we can merge this since we are breaking this.

The tests are fine, it's a bcachefs bug I think, I hit it myself, but it compiled fine on my XFS VPS, and @RaitoBezarius 's system.

Madouura reviewed (approved):
I am tentatively approving this since all bcachefs tests pass.
The cryptsetup tests issue, which IIRC is a util-linux issue, is a blocker however I believe.

Mic92 commented:
> I am tentatively approving this since all bcachefs tests pass. The cryptsetup tests issue, which IIRC is a util-linux issue, is a blocker however I believe.

If it is the fact that using bcachefs rather than the util-linux patch is the reason for the failing test, than it's not a blocker because it's a different issue and hydra doesn't use bcachefs.

Mic92 commented:
Can you rebase this towards staging?

YellowOnion committed:
bcachefs: Add YellowOnion as maintainer

YellowOnion committed:
bcachefs: enable ftrace for upstream bug reports

YellowOnion committed:
bcachefs: fix version not showing correctly

buildLinux doesn't provide argsOverride by default.

Prior to 663caaa, a default kernel was replaced with a direct invocation
to buildLinux, this broke the use of argsOverride, in 663caaa, the use
of argsOverride was removed, and then later the invocation of buildLinux
was replaced with vanilla linux kernel that required argsOverride
which then rebroke things again.

Recommend either: put argsOverride inside buildLinux, or replace it with
something less confusing to maintainers.

Mic92 commented on pkgs/os-specific/linux/kernel/bcachefs.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "diffHash": "sha256-BmXd/5Hn/xr7gNFp6BsBMG2XeKFlPGxv66IQ8DEwh5k=",
```
I got something else here:

```suggestion
  "diffHash": "sha256-DtMc8P4lTRzvS6PVvD7WtWEPsfnxIXSpqMsKKWs+edI=",
```

Mic92 commented on pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix:
```diff
@@ -2,40 +2,47 @@
 # but this does not means that it'll be the latest-compatible revision
-, currentCommit ? "84f132d5696138bb038d2dc8f1162d2fab5ac832"
-, diffHash ? "sha256-RaBWBU7rXjJFb1euFAFBHWCBQAG7npaCodjp/vMYpyw="
+, version ? lib.importJSON ./bcachefs.json
```
Given we might soon get rid of this kernel, it might be not worth it to have complex update logic here. Some hashes here might be enough.

Mic92 commented on pkgs/tools/filesystems/bcachefs-tools/default.nix:
```diff
@@ -12,31 +12,20 @@
 , rustPlatform
 , makeWrapper
 , fuseSupport ? false
+, version ? lib.importJSON ./version.json
```
How does one generate this file? without having automation i.e. `passthru.updateScript` in place its better to use put versions in json. It will break tools like nix-update and ryantm bot.

Mic92 committed:
bcachefs: revert using json to hold bcachefs commits/hashes

The json appearantly contained the wrong checksum and we don't have a way to generate this

Mic92 commented on pkgs/tools/filesystems/bcachefs-tools/version.json:
```diff
@@ -0,0 +1,7 @@
+  "owner": "koverstreet",
+  "repo": "bcachefs-tools",
+  "rev": "6b175a022496572416918bd38d083120c23ba5f2",
+  "sha256": "qC6Bq2zdO8Tj+bZbIUvcVBqvuKccqDEX3HIeOXsEloQ=",
```
This doesn't look like a valid checksum either.

Mic92 replied:
Ah this is the old style hash.

Mic92 commented on pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix:
```diff
@@ -2,40 +2,47 @@
-  };
+  argsOverride = {
+    version = "${kernel.version}${localversion}";
+    modDirVersion = "${kernel.version}${localversion}";
```
modDirversion must be of this format `\d+\.\d+\.\d+` you cannot put strings in it.

Mic92 commented:
I fixed the build locally and made sure all nixos tests for bcachefs run. I suggest to move these json refactorings to a later PR, when this is merged into master.

vcunat commented:
This conflicted with PR #265820.  Perhaps you want to check the result.

amjoseph-nixpkgs commented:
> Perhaps you want to check the result in [0fe5300](https://github.com/NixOS/nixpkgs/commit/0fe5300699182cd6e2d6f6e2459671e6d5ffd69b)

It's okay.

Also, I merged

- https://github.com/NixOS/nixpkgs/pull/267895

which bumps us to 1.3.3.

lukegb commented:
This broke bcache during install (nixosTests.installer.bcache); `grep(!qr#dev/bcachefs.*#, @bcacheDevices)` is equivalent to 
`grep(!true, @bcacheDevices)`, because the `qr` operator returns a regex operator, rather than matching against `$_`.
